cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,446 @@
+
+    private final String className;
+    private boolean interfaceClass;
+    private final Map<Method, AnnotationDetails> methodAnnotations;
+
+    AuthEnforceAnnotationVisitor(String className) {
+      super(Opcodes.ASM5);
+      this.className = className;
+      this.methodAnnotations = new HashMap<>();
+    }
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+      // visit the header of the class to identify if this class is an interface
+      super.visit(version, access, name, signature, superName, interfaces);
+      interfaceClass = Modifier.isInterface(access);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(final int access, final String methodName, final String methodDesc,
+                                     final String signature, final String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+
+      // No need to visit the class to look for AuthEnforce annotation if this class is an interface or the method is
+      // static.
+      if (interfaceClass || Modifier.isStatic(access)) {
+        return mv;
+      }
+
+
+      // Visit the annotations of the method to determine if it has AuthEnforce annotation. If it does then collect
+      // all the AuthEnforce annotation details through AnnotationVisitor and also sets a boolean flag hasEnforce
+      // which is used later in visitParameterAnnotation to visit the annotations on parameters of this method only
+      // if it had AuthEnforce annotation.
+      mv = new MethodVisitor(Opcodes.ASM5, mv) {
+
+        boolean hasAuthEnforce;
+        AuthAnnotation authAnnotation;
+        final Map<String, Integer> paramAnnotation = new HashMap<>();
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
+
+
+          // If the annotation is not visible or its not AuthEnforce just skip
+          if (!(visible && AuthEnforce.class.getName().equals(Type.getType(desc).getClassName()))) {
+            return annotationVisitor;
+          }
+
+          // AuthEnforce annotation is present so visit it
+          // set the flag which will be used to decide whether to visit parameter annotation or not
+          hasAuthEnforce = true;
+          final Type[] enforceOn = new Type[1];
+          final List<String> entities = new ArrayList<>();
+          final Set<Action> actions = new HashSet<>();
+
+          // We found AuthEnforce annotation so visit the annotation to collect all the specified details
+          // See AnnotationVisitor documentation to know the order in which the below overridden methods are
+          // called and their responsibilities.
+          annotationVisitor = new AnnotationVisitor(Opcodes.ASM5, annotationVisitor) {
```

Comment:
I think you can simplify this by using `AnnotationNode` instead. Do we support multiple `@AuthEnforce` annotations on the same method? If not, then it can simply be:

```java
new MethodVisitor(...) {
  AnnotationNode authEnforceNode;   // if we support multiple @AuthEnforce, then this becomes a List<>
  
  @Override
  public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
    if (!(visible && AuthEnforce.class.getName().equals(Type.getType(desc).getClassName()))) {
     return super.visitAnnotation(desc, visible);
    }
    authEnforceNode = new AnnotationNode(desc);
    return authEnforceNode;
  }
}
```